In [20]:
import pandas as pd
ticks = pd.read_csv ('ticks.csv', sep="|")
ticks['gvkey'] = ticks['gvkey'].fillna(0.0).astype(int)
ticks['Permco'] = ticks['Permco'].fillna(0.0).astype(int)
ticks

Primary Dealer  \
0                   ABN AMRO INCORPORATED   
1           AUBREY G. LANSTON & CO., INC.   
2                           BA Securities   
3                                Banc One   
4                         Bank of America   
..                                    ...   
107         WERTHEIM SCHRODER & CO., INC.   
108  WESTPAC POLLOCK GOV'T SECURITIES INC   
109                 WHITE, WELD & CO INC.   
110        YAMAICHI INT'L (AMERICA), INC.   
111                                 Zions   

                                       Holding Company         Ticker  \
0                                                  NaN            ABN   
1                                       Citigroup Inc.       C (NYSE)   
2                          Bank of America Corporation     BAC (NYSE)   
3    JPMorgan Chase & Co. (Acquired by JPMorgan Chase)     JPM (NYSE)   
4                          Bank of America Corporation     BAC (NYSE)   
..                                                 ...            ...   
107                       Schroder Wertheim & Co. Inc.            NaN   
108                        Westpac Banking Corporation     WBK (NYSE)   
109                          Merrill Lynch & Co., Inc.            NaN   
110                Yamaichi Securities Company Limited            NaN   
111                               Zions Bancorporation  ZION (NASDAQ)   

     Start Date   End Date  Permco  gvkey  
0     9/29/1998  9/15/2006   31989  15504  
1     5/19/1960  4/17/2000   20483   3243  
2     4/18/1994  9/30/1997     437   7647  
3      4/1/1999   8/1/2004   20436   2968  
4    11/17/1971  4/15/1994     437   7647  
..          ...        ...     ...    ...  
107   6/24/1988  11/8/1990       0      0  
108    2/4/1987  6/27/1990   22027  15362  
109   2/26/1976  4/18/1978   21190   7267  
110   9/29/1988  12/4/1997       0      0  
111   8/11/1993  3/31/2002   21305   8007  

[112 rows x 7 columns]

In [12]:
import pandas as pd
import wrds
import config
from datetime import datetime

db = wrds.Connection(wrds_username=config.WRDS_USERNAME)

Loading library list...
Done


In [56]:
def fetch_financial_data(pgvkey, start_date, end_date):
    """
    Fetch financial data for a given ticker and date range from the CCM database in WRDS.
    
    :param gvkey: The gvkey symbol for the company.
    :param start_date: The start date for the data in YYYY-MM-DD format.
    :param end_date: The end date for the data in YYYY-MM-DD format or 'Current'.
    :return: A DataFrame containing the financial data.
    """
    if not gvkey:  # Skip if no ticker is available
        return pd.DataFrame()
    
    # If the end date is 'Current', replace it with today's date
    if end_date == 'Current':
        end_date = datetime.today().strftime('%Y-%m-%d')
    
    # Convert dates to the correct format if necessary
    start_date = pd.to_datetime(start_date).strftime('%Y-%m-%d')
    end_date = pd.to_datetime(end_date).strftime('%Y-%m-%d')
    
        
    query = f"""
    SELECT datadate, at AS total_assets, lt AS book_debt, ceq AS book_equity, csho*prcc_f AS market_equity, gvkey, conm
    FROM comp.funda as cst
    WHERE cst.gvkey = '{str(pgvkey).zfill(6)}'
    AND cst.datadate BETWEEN '{start_date}' AND '{end_date}'
    AND indfmt='INDL'
    AND datafmt='STD'
    AND popsrc='D'
    AND consol='C'
    """
    data = db.raw_sql(query)
    return data

Get all data for primary dealers

In [57]:
empty_tickers = []
prim_dealers = pd.DataFrame()
# Iterate over DataFrame rows and fetch data for each ticker
for index, row in ticks.iterrows():
    gvkey = row['gvkey']
    start_date = row['Start Date']
    end_date = row['End Date']      # Formatting date for the query
    
    # Fetch financial data for the ticker if available
    new_data = fetch_financial_data(gvkey, start_date, end_date)
    if isinstance(new_data, tuple):
        empty_tickers.append({row['Ticker']:gvkey})
    else:
        prim_dealers = pd.concat([new_data, prim_dealers], axis=0)
print(empty_tickers)

C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\3471547033.py:14: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  prim_dealers = pd.concat([new_data, prim_dealers], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\3471547033.py:14: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  prim_dealers = pd.concat([new_data, prim_dealers], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\3471547033.py:14: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is

[]


Get data from all firms in Compustat

In [59]:
prim_dealers = prim_dealers.drop_duplicates()
prim_dealers

datadate  total_assets    book_debt  book_equity  market_equity   gvkey  \
0   1993-12-31     50782.297    47213.898     3226.500    7121.765625  008007   
1   1994-12-31     59315.898    55469.500     3334.400    7226.264375  008007   
2   1995-12-31     72134.375    66822.250     5009.797   11641.080000  008007   
3   1996-12-31     80175.375    74111.188     5875.398   16038.580500  008007   
4   1997-12-31     88540.188    81518.000     6834.199   29396.486250  008007   
..         ...           ...          ...          ...            ...     ...   
3   2001-12-31    531713.000   517166.000     9767.000   24997.940000  015504   
4   2002-12-31    582984.873   567686.209    10450.400   25655.008000  015504   
5   2003-12-31    705982.489   684869.917    15411.170   38457.492840  015504   
6   2004-12-31    823953.817   797891.813    19230.729   44400.773200  015504   
7   2005-12-31   1043048.097  1014447.298    26314.109   49088.828800  015504   

                    conm  
0       WELLS FARGO & CO  
1       WELLS FARGO & CO  
2       WELLS FARGO & CO  
3       WELLS FARGO & CO  
4       WELLS FARGO & CO  
..                   ...  
3   ABN-AMRO HOLDINGS NV  
4   ABN-AMRO HOLDINGS NV  
5   ABN-AMRO HOLDINGS NV  
6   ABN-AMRO HOLDINGS NV  
7   ABN-AMRO HOLDINGS NV  

[816 rows x 7 columns]

In [61]:
yearly_sum_pd = prim_dealers.groupby('datadate').agg({
    'total_assets': 'sum',
    'book_debt': 'sum',
    'book_equity': 'sum',
    'market_equity': 'sum'
}).reset_index()
yearly_sum_pd

datadate  total_assets     book_debt  book_equity  market_equity
0    1960-12-31  8.835297e+03  4.505152e+03      746.347   0.000000e+00
1    1961-12-31  9.856233e+03  5.006815e+03      771.122   8.476998e+02
2    1962-12-31  1.039177e+04  5.191507e+03      811.380   1.936547e+03
3    1963-12-31  1.135703e+04  9.798979e+03     1377.017   1.962843e+03
4    1964-12-31  1.885544e+04  1.670331e+04     1818.247   2.841011e+03
..          ...           ...           ...          ...            ...
153  2022-12-31  1.966493e+07  1.821493e+07  1321360.741   1.257698e+06
154  2023-03-31  2.271969e+06  2.178447e+06    23685.000   4.714652e+04
155  2023-10-31  6.666081e+06  6.280510e+06   361929.000   4.375408e+05
156  2023-11-30  5.790516e+04  4.810262e+04     9709.785   7.464621e+03
157  2023-12-31  8.776014e+06  8.067776e+06   651893.083   8.579271e+05

[158 rows x 5 columns]

In [5]:
def pull_CRSP_Comp_Link_Table(wrds_username=config.WRDS_USERNAME):
    sql_query = """
        SELECT 
            gvkey, lpermco AS permco, linktype, linkprim, linkdt, linkenddt, tic
        FROM 
            crsp.ccmxpf_linkhist
        WHERE 
            substr(linktype,1,1)='L' AND 
            (linkprim ='C' OR linkprim='P')
        """
    db = wrds.Connection(wrds_username=wrds_username)
    ccm = db.raw_sql(sql_query, date_cols=["linkdt", "linkenddt"])
    db.close()
    return ccm

In [ ]:
pull_CRSP_Comp_Link_Table()

In [35]:
linktable = pd.read_csv('linktable.csv')
linktable

gvkey                       conm     tic     sic LINKPRIM LIID  \
0        1000      A & E PLASTIK PAK INC    AE.2  3089.0        P   01   
1        1001    A & M FOOD SERVICES INC   AMFD.  5812.0        P   01   
2        1002                   AAI CORP  AAIC.1  3825.0        C   01   
3        1003      A.A. IMPORTING CO INC    ANTQ  5712.0        C   01   
4        1004                   AAR CORP     AIR  5080.0        P   01   
...       ...                        ...     ...     ...      ...  ...   
31944  349994     CLEARMIND MEDICINE INC    CMND  2834.0        P   01   
31945  350681       GETNET ADQUIRENCIA E     GET  7374.0        P   90   
31946  351038  QUOIN PHARMACEUTICALS LTD    QNRX  2834.0        P   90   
31947  352262           COOL COMPANY LTD    CLCO  4400.0        P   01   
31948  353444                 HALEON PLC     HLN  2834.0        P   90   

      LINKTYPE  LPERMNO  LPERMCO      LINKDT  LINKENDDT  
0           LU    25881    23369  11/13/1970  6/30/1978  
1           LU    10015     6398   9/20/1983  7/31/1986  
2           LC    10023    22159  12/14/1972   6/5/1973  
3           LU    10031     6672   12/7/1983  8/16/1989  
4           LU    54594    20000   4/24/1972          E  
...        ...      ...      ...         ...        ...  
31944       LC    23514    59438  11/15/2022          E  
31945       LC    22205    58855  10/22/2021  3/31/2023  
31946       LC    16161    55612  10/29/2021          E  
31947       LC    23773    59507   3/17/2023          E  
31948       LC    23209    59330   7/22/2022          E  

[31949 rows x 11 columns]

In [36]:
merged = pd.merge(ticks, linktable, left_on='gvkey', right_on='gvkey')
merged

Primary Dealer  \
0                  ABN AMRO INCORPORATED   
1          AUBREY G. LANSTON & CO., INC.   
2                              Citigroup   
3                          BA Securities   
4                        Bank of America   
..                                   ...   
93  SMITH BARNEY, HARRIS UPHAM & CO.,INC   
94        SECURITY PACIFIC NATIONAL BANK   
95              SOUTHERN CALIF S&L ASSOC   
96               TD SECURITIES (USA) LLC   
97                                 Zions   

                                    Holding Company          Ticker  \
0                                               NaN             ABN   
1                                    Citigroup Inc.        C (NYSE)   
2                                    Citigroup Inc.        C (NYSE)   
3                       Bank of America Corporation      BAC (NYSE)   
4                       Bank of America Corporation      BAC (NYSE)   
..                                              ...             ...   
93                            Primerica Corporation             PRI   
94                     Security Pacific Corporation            6020   
95  California Federal Savings and Loan Association             NaN   
96                        The Toronto-Dominion Bank  TD (NYSE, TSX)   
97                             Zions Bancorporation   ZION (NASDAQ)   

    Start Date   End Date  Permco  gvkey                        conm    tic  \
0    9/29/1998  9/15/2006   31989  15504        ABN-AMRO HOLDINGS NV  ABNYY   
1    5/19/1960  4/17/2000   20483   3243               CITIGROUP INC      C   
2    6/15/1961    Current   20483   3243               CITIGROUP INC      C   
3    4/18/1994  9/30/1997     437   7647        BANK OF AMERICA CORP    BAC   
4   11/17/1971  4/15/1994     437   7647        BANK OF AMERICA CORP    BAC   
..         ...        ...     ...    ...                         ...    ...   
93   8/22/1979  8/31/1998   22177   1414               PRIMERICA INC    PRI   
94  12/11/1986  1/17/1991    4205   9577       SECURITY PACIFIC CORP  7448B   
95    6/7/1983   8/5/1983    5884   2612  CALIFORNIA FED BANCORP INC  CAL.2   
96   2/11/2014    Current   29152  15706       TORONTO DOMINION BANK     TD   
97   8/11/1993  3/31/2002   21305   8007            WELLS FARGO & CO    WFC   

       sic LINKPRIM LIID LINKTYPE  LPERMNO  LPERMCO      LINKDT  LINKENDDT  
0   6020.0        P   90       LC    84766    31989   5/21/1997  4/24/2008  
1   6199.0        P   01       LC    70519    20483  10/29/1986          E  
2   6199.0        P   01       LC    70519    20483  10/29/1986          E  
3   6020.0        P   01       LC    59408     3151  12/14/1972          E  
4   6020.0        P   01       LC    59408     3151  12/14/1972          E  
..     ...      ...  ...      ...      ...      ...         ...        ...  
93  6311.0        C  00X       LC    10241    22177    1/1/1950  1/30/1962  
94  6020.0        P   01       LU    60839     4205  12/14/1972  4/30/1992  
95  6035.0        P   01       LC    66456     5884   3/29/1983   1/3/1997  
96  6020.0        P   01       LC    83835    29152   8/30/1996          E  
97  6020.0        P   01       LC    38703    21305  12/10/1962          E  

[98 rows x 17 columns]

### Get all data from broker dealers with SIC code 6211 or 6221 that are not in list of primary dealers


In [103]:
linked_bd = linktable[((linktable['sic'] == 6211) | (linktable['sic'] == 6221)) & (~linktable['gvkey'].isin(ticks['gvkey'].tolist()))]
linked_bd

gvkey                          conm    tic     sic LINKPRIM LIID  \
167      1148              ADVEST GROUP INC  ADV.2  6211.0        P   01   
987      1967               BACHE GROUP INC   BAC.  6211.0        P   01   
1379     2330          BOSTON CO INC  -CL B  BSTNB  6211.0        C  00X   
1380     2330          BOSTON CO INC  -CL B  BSTNB  6211.0        C   01   
1475     2432    BROWN (ROBERT C.) & CO INC   RCBI  6211.0        P   01   
...       ...                           ...    ...     ...      ...  ...   
30429  178073        B. RILEY FINANCIAL INC   RILY  6211.0        P   03   
30479  178519    LPL FINANCIAL HOLDINGS INC   LPLA  6211.0        P   01   
30548  178848  DIRECT MARKETS HOLDINGS CORP  MKTSQ  6211.0        P   01   
31100  185518          GLOBAL BROKERAGE INC   GLBR  6211.0        C   01   
31489  209386             VAN DER MOOLEN NV  VDMHY  6211.0        P   90   

      LINKTYPE  LPERMNO  LPERMCO      LINKDT   LINKENDDT  
167         LU    64514      420  11/21/1980   2/28/2001  
987         LU    54893    25401   5/31/1972   6/11/1981  
1379        LU    18877    22881    1/1/1975   1/30/1975  
1380        LU    18877    22881   1/31/1975   9/30/1981  
1475        LU    19538     6441   9/30/1983  11/30/1989  
...        ...      ...      ...         ...         ...  
30429       LC    15540    55369   7/16/2015           E  
30479       LC    12400    53576  11/18/2010           E  
30548       LC    92312    52781  10/16/2007   9/13/2012  
31100       LC    12446    53607   12/2/2010           E  
31489       LC    89197    42289  10/18/2001   12/5/2007  

[165 rows x 11 columns]

In [105]:
empty_tickers = []
broker_dealers = pd.DataFrame()
for index, row in linked_bd.iterrows():
    gvkey = row['gvkey']
    start_date = '1960-01-01'
    end_date = '2012-12-31'    
    new_data = fetch_financial_data(gvkey, start_date, end_date)
    if isinstance(new_data, tuple):
        empty_tickers.append({row['conm']:gvkey})
    else:
        broker_dealers = pd.concat([new_data, broker_dealers], axis=0)
print(empty_tickers)

C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\1247283076.py:11: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  broker_dealers = pd.concat([new_data, broker_dealers], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\1247283076.py:11: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  broker_dealers = pd.concat([new_data, broker_dealers], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\1247283076.py:11: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA en

[]


In [106]:
broker_dealers = broker_dealers.drop_duplicates()
broker_dealers

datadate  total_assets  book_debt  book_equity  market_equity   gvkey  \
0   1999-12-31       842.252    729.293          NaN            NaN  209386   
1   2000-12-31       876.311    726.460          NaN            NaN  209386   
2   2001-12-31      3631.252   3064.614      538.155    1098.808600  209386   
3   2002-12-31      1567.193   1174.844      327.342     825.624310  209386   
4   2003-12-31      1114.079    818.679      253.830     333.861110  209386   
..         ...           ...        ...          ...            ...     ...   
15  1996-09-30       965.177    875.935       89.242      81.929250  001148   
16  1997-09-30      1078.839    973.186      105.653     227.151465  001148   
17  1998-09-30      1233.271   1109.604      123.667     182.335875  001148   
18  1999-09-30      1462.621   1327.485      135.136     162.881250  001148   
19  2000-09-30      2088.889   1938.089      150.800     279.927375  001148   

                 conm  
0   VAN DER MOOLEN NV  
1   VAN DER MOOLEN NV  
2   VAN DER MOOLEN NV  
3   VAN DER MOOLEN NV  
4   VAN DER MOOLEN NV  
..                ...  
15   ADVEST GROUP INC  
16   ADVEST GROUP INC  
17   ADVEST GROUP INC  
18   ADVEST GROUP INC  
19   ADVEST GROUP INC  

[1627 rows x 7 columns]

In [107]:
yearly_sum_bd = broker_dealers.groupby('datadate').agg({
    'total_assets': 'sum',
    'book_debt': 'sum',
    'book_equity': 'sum',
    'market_equity': 'sum'
}).reset_index()
yearly_sum_bd = yearly_sum_bd.rename(columns={'total_assets':'total_assets_bd', 'book_debt':'book_debt_bd', 'book_equity':'book_equity_bd','market_equity':'market_equity_bd' })
yearly_sum_bd

datadate  total_assets_bd  book_debt_bd  book_equity_bd  \
0    1960-12-31           64.700        30.700           0.000   
1    1961-12-31           75.200        39.984           0.000   
2    1962-12-31          596.288       536.342          59.895   
3    1963-12-31          624.431       561.682          23.477   
4    1964-12-31          716.440       649.862          23.733   
..          ...              ...           ...             ...   
366  2012-03-31           15.491         8.235           7.256   
367  2012-06-30         3561.298      3203.698         357.600   
368  2012-09-30        40689.854     32587.463        7684.875   
369  2012-11-30        36293.541     32162.737        3436.015   
370  2012-12-31       250033.509    222541.763       21890.761   

     market_equity_bd  
0            0.000000  
1            0.000000  
2            0.000000  
3            0.000000  
4            0.000000  
..                ...  
366         26.673150  
367        175.774794  
368      13429.889470  
369       3448.171520  
370      24642.680700  

[371 rows x 5 columns]

In [108]:
merged_pd_bd = pd.merge(yearly_sum_pd, yearly_sum_bd, left_on='datadate', right_on='datadate')
merged_pd_bd.index = pd.to_datetime(merged_pd_bd['datadate'])
merged_pd_bd['year'] = pd.to_datetime(merged_pd_bd['datadate']).dt.year
merged_pd_bd = merged_pd_bd.groupby('year').agg({
    'total_assets': 'sum',
    'book_debt': 'sum',
    'book_equity': 'sum',
    'market_equity': 'sum',
    'total_assets_bd': 'sum',
    'book_debt_bd': 'sum',
    'book_equity_bd': 'sum',
    'market_equity_bd': 'sum'
})
merged_pd_bd

total_assets     book_debt  book_equity  market_equity  total_assets_bd  \
year                                                                            
1960  8.835297e+03  4.505152e+03      746.347   0.000000e+00           64.700   
1961  9.856233e+03  5.006815e+03      771.122   8.476998e+02           75.200   
1962  1.039177e+04  5.191507e+03      811.380   1.936547e+03          596.288   
1963  1.135703e+04  9.798979e+03     1377.017   1.962843e+03          624.431   
1964  1.885544e+04  1.670331e+04     1818.247   2.841011e+03          716.440   
1965  2.057146e+04  1.826910e+04     1920.637   2.882466e+03          692.547   
1966  2.191825e+04  1.952541e+04     2033.901   2.583758e+03          877.911   
1967  2.458797e+04  2.208529e+04     2122.413   2.702707e+03         1331.803   
1968  2.771266e+04  2.507688e+04     2225.453   3.620185e+03         1145.109   
1969  3.246661e+04  2.952803e+04     2469.511   3.213643e+03         1422.050   
1970  3.834176e+04  3.496507e+04     2892.354   3.339911e+03         2208.451   
1971  4.576397e+04  4.202574e+04     3229.665   4.482702e+03         2487.608   
1972  5.593672e+04  5.196867e+04     3420.911   4.906063e+03         3689.163   
1973  7.021171e+04  6.586432e+04     3740.116   3.970551e+03         3657.981   
1974  7.848777e+04  7.372397e+04     3996.323   2.592040e+03         4275.996   
1975  8.210193e+04  7.748725e+04     4499.748   3.092486e+03         6707.453   
1976  9.132380e+04  8.630369e+04     4900.147   4.679208e+03         9769.746   
1977  1.079246e+05  1.023092e+05     5401.737   4.317002e+03        11395.891   
1978  1.275664e+05  1.207351e+05     6644.098   4.471928e+03        10596.699   
1979  1.669253e+05  1.518684e+05    14717.089   1.659735e+04        14634.381   
1980  1.868499e+05  1.699945e+05    16478.633   2.036287e+04        17319.795   
1981  2.098301e+05  1.909973e+05    18485.997   2.021323e+04        23175.286   
1982  2.243988e+05  2.035529e+05    20275.462   2.957431e+04        65170.982   
1983  3.334537e+05  3.057589e+05    26269.035   3.952904e+04        73210.811   
1984  4.410808e+05  4.071473e+05    31166.377   3.871165e+04        89288.325   
1985  4.985783e+05  4.608472e+05    34865.781   4.984015e+04       144384.903   
1986  6.287239e+05  5.817710e+05    44064.969   6.898620e+04       196188.834   
1987  7.426556e+05  6.910977e+05    47521.715   6.464965e+04       192707.141   
1988  8.148282e+05  7.544517e+05    55703.189   6.761058e+04       186116.133   
1989  1.123925e+06  1.048617e+06    69967.406   1.085505e+05       193806.556   
1990  1.164147e+06  1.087964e+06    69959.988   8.550142e+04       185882.819   
1991  1.137509e+06  1.064395e+06    66850.626   1.171588e+05       254364.028   
1992  1.090730e+06  1.017253e+06    67647.809   1.321120e+05       336575.724   
1993  1.854017e+06  1.739590e+06    94962.858   2.051288e+05       299438.369   
1994  1.794131e+06  1.710128e+06    76825.838   1.055642e+05       315795.953   
1995  2.196213e+06  2.082875e+06   104712.486   1.591099e+05       364611.139   
1996  2.594022e+06  2.468386e+06   116163.679   2.189292e+05       397384.235   
1997  3.552339e+06  3.392502e+06   146848.886   3.744262e+05       511448.293   
1998  5.531776e+06  5.250772e+06   258206.152   6.201915e+05       460574.457   
1999  6.116472e+06  5.796696e+06   297307.631   7.829493e+05       563039.502   
2000  8.800216e+06  8.328568e+06   439939.052   1.218901e+06       759439.939   
2001  8.658743e+06  8.181125e+06   452506.063   1.073983e+06       748709.621   
2002  8.574401e+06  8.126400e+06   428138.680   7.732833e+05       756620.599   
2003  1.020420e+07  9.677706e+06   498795.914   1.077202e+06       955024.169   
2004  1.306258e+07  1.236090e+07   664322.692   1.286215e+06      1125717.886   
2005  1.619906e+07  1.543661e+07   691025.489   1.305207e+06       784551.122   
2006  1.783142e+07  1.696300e+07   796219.358   1.571207e+06       591575.942   
2007  2.188721e+07  2.094492e+07   81

In [110]:
merged_pd_bd['tot_assets_rat'] = merged_pd_bd['total_assets']/(merged_pd_bd['total_assets_bd']+merged_pd_bd['total_assets'])
merged_pd_bd['book_debt_rat'] = merged_pd_bd['book_debt']/(merged_pd_bd['book_debt_bd']+merged_pd_bd['book_debt'])
merged_pd_bd['book_equity_rat'] = merged_pd_bd['book_equity']/(merged_pd_bd['book_equity_bd']+merged_pd_bd['book_equity'])
merged_pd_bd['market_equity_rat'] = merged_pd_bd['market_equity']/(merged_pd_bd['market_equity_bd']+merged_pd_bd['market_equity'])
merged_pd_bd[['tot_assets_rat', 'book_debt_rat', 'book_equity_rat','market_equity_rat']].mean()

tot_assets_rat       0.909538
book_debt_rat        0.908265
book_equity_rat      0.919713
market_equity_rat    0.912978
dtype: float64

Get data from banks
6011 - Federal Reserve Banks
6021 - National Commercial Banks
6022 - State Commercial Banks
6029 - Commercial Banks, nec (not elsewhere classified)
6081 - Foreign Bank and Branches and Agencies
6082 - Foreign Trade and International Banking Institutions

In [153]:
bank_sic = [6011,6021,6022,6029,6081,6082,6020] # these are not all the codes, in the process of looking these up
linked_banks = linktable[(linktable['sic'].isin(bank_sic)) & (~linktable['gvkey'].isin(ticks['gvkey'].tolist()))]
linked_banks.drop_duplicates()

gvkey                          conm    tic     sic LINKPRIM LIID  \
197      1178    AFFILIATED BANKSHARES COLO  AFBK.  6020.0        C  00X   
198      1178    AFFILIATED BANKSHARES COLO  AFBK.  6020.0        P   01   
309      1290      ALLIED BANCSHARES INC-TX   ALBN  6020.0        P   01   
310      1290      ALLIED BANCSHARES INC-TX   ALBN  6020.0        C  00X   
485      1458        AMERICAN FLETCHER CORP   AFLT  6020.0        C   01   
...       ...                           ...    ...     ...      ...  ...   
31673  260758    ATLANTIC SOUTHERN FINL GRP   ASFN  6020.0        C   01   
31703  264418    1ST CENTURY BANCSHARES INC   FCTY  6020.0        P   01   
31721  266160    CITIZENS COMMUNITY BANCORP   CZWI  6020.0        P   01   
31807  285313  BANCO SANTANDER BRASIL  -ADR   BSBR  6020.0        P   90   
31808  285333  INTERCORP FINANCIAL SVCS INC    IFS  6020.0        P   02   

      LINKTYPE  LPERMNO  LPERMCO      LINKDT   LINKENDDT  
197         LU    11173       94    1/1/1974   1/30/1974  
198         LU    11173       94   1/31/1974  11/30/1992  
309         LC    11870      151   1/31/1974   1/29/1988  
310         LC    11870      151    1/1/1974   1/30/1974  
485         LU    12969      106  12/14/1972   1/23/1987  
...        ...      ...      ...         ...         ...  
31673       LC    91955    52130   4/16/2007   1/28/2011  
31703       LC    92870    53112  12/18/2008   6/30/2016  
31721       LC    91609    51445   11/1/2006           E  
31807       LC    93054    53230   10/7/2009           E  
31808       LC    18828    56820   7/19/2019           E  

[1414 rows x 11 columns]

In [144]:
def get_comparison_group_data(linktable_df, start_date, end_date):
    empty_tickers = []
    comparison_group_df = pd.DataFrame()
    for index, row in linktable_df.iterrows():
        gvkey = row['gvkey']   
        new_data = fetch_financial_data(gvkey, start_date, end_date)
        if isinstance(new_data, tuple):
            empty_tickers.append({row['conm']:gvkey})
        else:
            comparison_group_df = pd.concat([new_data, comparison_group_df], axis=0)
    print('Empty Tickers: %s' % empty_tickers)
    return comparison_group_df

In [145]:
banks = get_comparison_group_data(linked_banks, '1960-01-01', '2012-12-31')
banks = banks.drop_duplicates()

C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\2771333044.py:10: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  comparison_group_df = pd.concat([new_data, comparison_group_df], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\2771333044.py:10: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  comparison_group_df = pd.concat([new_data, comparison_group_df], axis=0)
C:\Users\jsime\AppData\Local\Temp\ipykernel_19320\2771333044.py:10: FutureWarning: The behavior of DataFrame concatenation wit

Empty Tickers: []


In [146]:
banks

datadate  total_assets   book_debt  book_equity  market_equity   gvkey  \
0   2007-12-31     60887.708   55149.327     5738.349            NaN  285313   
1   2008-12-31    126915.378  105415.604    21497.497            NaN  285313   
2   2009-12-31    181332.899  141581.540    39750.591   52929.469060  285313   
3   2010-12-31    225279.708  181167.271    44107.581   51638.506400  285313   
4   2011-12-31    214680.884  172789.011    41881.694   30876.314260  285313   
..         ...           ...         ...          ...            ...     ...   
13  1987-12-31      2536.776    2354.405      171.473     121.224000  001178   
14  1988-12-31      2616.952    2439.173      167.005      97.241375  001178   
15  1989-12-31      2611.094    2425.620      174.703     122.208000  001178   
16  1990-12-31      2738.097    2545.413      181.912     109.897250  001178   
17  1991-12-31      2912.117    2702.303      199.207     326.781000  001178   

                            conm  
0   BANCO SANTANDER BRASIL  -ADR  
1   BANCO SANTANDER BRASIL  -ADR  
2   BANCO SANTANDER BRASIL  -ADR  
3   BANCO SANTANDER BRASIL  -ADR  
4   BANCO SANTANDER BRASIL  -ADR  
..                           ...  
13    AFFILIATED BANKSHARES COLO  
14    AFFILIATED BANKSHARES COLO  
15    AFFILIATED BANKSHARES COLO  
16    AFFILIATED BANKSHARES COLO  
17    AFFILIATED BANKSHARES COLO  

[17052 rows x 7 columns]

In [152]:
linked_banks['sic'].value_counts()

sic
6020.0    1413
6022.0       1
Name: count, dtype: int64

In [147]:
yearly_sum_banks = banks.groupby('datadate').agg({
    'total_assets': 'sum',
    'book_debt': 'sum',
    'book_equity': 'sum',
    'market_equity': 'sum'
}).reset_index()
yearly_sum_banks = yearly_sum_banks.rename(columns={'total_assets':'total_assets_bd', 'book_debt':'book_debt_bd', 'book_equity':'book_equity_bd','market_equity':'market_equity_bd' })
yearly_sum_banks

datadate  total_assets_bd  book_debt_bd  book_equity_bd  \
0    1960-12-31        78599.073     41552.276        6361.886   
1    1961-12-31        86176.330     43121.544        6858.341   
2    1962-12-31        92296.289     48642.078        7224.936   
3    1963-12-31       126363.247    115456.199        9614.007   
4    1964-12-31       140929.862    129064.851       10389.785   
..          ...              ...           ...             ...   
165  2012-03-31      4534458.104   4320424.337      180131.325   
166  2012-06-30         4502.981      4026.934         476.043   
167  2012-09-30      1489418.503   1396915.511       86393.809   
168  2012-10-31         1534.168      1441.064          93.104   
169  2012-12-31      8624042.753   7800908.266      735181.372   

     market_equity_bd  
0            0.000000  
1            0.000000  
2        10988.304739  
3        12355.733646  
4        13035.880255  
..                ...  
165     162161.210470  
166        221.196470  
167     131908.920423  
168          0.000000  
169     744948.192681  

[170 rows x 5 columns]

In [148]:
merged_pd_banks = pd.merge(yearly_sum_pd, yearly_sum_banks, left_on='datadate', right_on='datadate')
merged_pd_banks.index = pd.to_datetime(merged_pd_banks['datadate'])
merged_pd_banks['year'] = pd.to_datetime(merged_pd_banks['datadate']).dt.year
merged_pd_banks = merged_pd_banks.groupby('year').agg({
    'total_assets': 'sum',
    'book_debt': 'sum',
    'book_equity': 'sum',
    'market_equity': 'sum',
    'total_assets_bd': 'sum',
    'book_debt_bd': 'sum',
    'book_equity_bd': 'sum',
    'market_equity_bd': 'sum'
})
merged_pd_banks

total_assets     book_debt  book_equity  market_equity  total_assets_bd  \
year                                                                            
1960  8.835297e+03  4.505152e+03      746.347   0.000000e+00     7.859907e+04   
1961  9.856233e+03  5.006815e+03      771.122   8.476998e+02     8.617633e+04   
1962  1.039177e+04  5.191507e+03      811.380   1.936547e+03     9.229629e+04   
1963  1.135703e+04  9.798979e+03     1377.017   1.962843e+03     1.263632e+05   
1964  1.885544e+04  1.670331e+04     1818.247   2.841011e+03     1.409299e+05   
1965  2.057146e+04  1.826910e+04     1920.637   2.882466e+03     1.574110e+05   
1966  2.191825e+04  1.952541e+04     2033.901   2.583758e+03     1.718237e+05   
1967  2.458797e+04  2.208529e+04     2122.413   2.702707e+03     1.987475e+05   
1968  2.771266e+04  2.507688e+04     2225.453   3.620185e+03     2.244083e+05   
1969  3.246661e+04  2.952803e+04     2469.511   3.213643e+03     2.493724e+05   
1970  3.834176e+04  3.496507e+04     2892.354   3.339911e+03     2.795517e+05   
1971  4.576397e+04  4.202574e+04     3229.665   4.482702e+03     3.231808e+05   
1972  5.536908e+04  5.146461e+04     3384.766   4.862431e+03     3.893781e+05   
1973  7.021171e+04  6.586432e+04     3740.116   3.970551e+03     4.742442e+05   
1974  7.848777e+04  7.372397e+04     3996.323   2.592040e+03     5.377937e+05   
1975  8.157608e+04  7.702992e+04     4458.659   3.062454e+03     5.445676e+05   
1976  9.030359e+04  8.536055e+04     4850.514   4.639833e+03     5.894864e+05   
1977  1.063348e+05  1.008022e+05     5346.312   4.284289e+03     6.698816e+05   
1978  1.249964e+05  1.182569e+05     6579.715   4.414297e+03     7.944357e+05   
1979  1.642534e+05  1.492979e+05    14643.052   1.653838e+04     8.959792e+05   
1980  1.833495e+05  1.666100e+05    16396.739   2.023590e+04     9.902394e+05   
1981  2.066243e+05  1.879342e+05    18361.552   2.007643e+04     1.087751e+06   
1982  2.201037e+05  1.994307e+05    20120.893   2.935261e+04     1.232060e+06   
1983  3.334537e+05  3.057589e+05    26269.035   3.952904e+04     1.308789e+06   
1984  4.410808e+05  4.071473e+05    31166.377   3.871165e+04     1.418346e+06   
1985  4.985783e+05  4.608472e+05    34865.781   4.984015e+04     1.556528e+06   
1986  6.287239e+05  5.817710e+05    44064.969   6.898620e+04     1.795500e+06   
1987  7.426556e+05  6.910977e+05    47521.715   6.464965e+04     1.843022e+06   
1988  8.148282e+05  7.544517e+05    55703.189   6.761058e+04     2.021294e+06   
1989  1.123925e+06  1.048617e+06    69967.406   1.085505e+05     2.160324e+06   
1990  1.164147e+06  1.087964e+06    69959.988   8.550142e+04     2.209809e+06   
1991  1.137509e+06  1.064395e+06    66850.626   1.171588e+05     2.282740e+06   
1992  1.090730e+06  1.017253e+06    67647.809   1.321120e+05     2.549764e+06   
1993  1.773664e+06  1.662671e+06    92033.404   2.009039e+05     2.931808e+06   
1994  1.586942e+06  1.510803e+06    70380.838   9.809770e+04     3.244155e+06   
1995  2.079519e+06  1.970736e+06   100876.486   1.545827e+05     3.409911e+06   
1996  2.225149e+06  2.116073e+06   101764.679   1.969169e+05     3.704481e+06   
1997  3.106512e+06  2.966102e+06   130490.886   3.414489e+05     3.993006e+06   
1998  5.027756e+06  4.769243e+06   240381.152   5.907703e+05     4.030230e+06   
1999  6.116472e+06  5.796696e+06   297307.631   7.829493e+05     4.495541e+06   
2000  7.357844e+06  6.954087e+06   377805.052   1.058149e+06     5.642915e+06   
2001  7.616081e+06  7.185869e+06   406114.063   9.529482e+05     5.848453e+06   
2002  7.428992e+06  7.030821e+06   378995.680   6.714276e+05     7.112361e+06   
2003  8.885499e+06  8.419957e+06   440163.914   9.503177e+05     8.952065e+06   
2004  1.139862e+07  1.076695e+07   597462.692   1.156325e+06     8.908568e+06   
2005  1.418367e+07  1.349836e+07   619892.489   1.152371e+06     8.507563e+06   
2006  1.536903e+07  1.459571e+07   711173.358   1.364352e+06     9.639986e+06   
2007  1.903094e+07  1.819247e+07   72

In [151]:
merged_pd_banks['tot_assets_rat'] = merged_pd_banks['total_assets']/(merged_pd_banks['total_assets_bd']+merged_pd_banks['total_assets'])
merged_pd_banks['book_debt_rat'] = merged_pd_banks['book_debt']/(merged_pd_banks['book_debt_bd']+merged_pd_banks['book_debt'])
merged_pd_banks['book_equity_rat'] = merged_pd_banks['book_equity']/(merged_pd_banks['book_equity_bd']+merged_pd_banks['book_equity'])
merged_pd_banks['market_equity_rat'] = merged_pd_banks['market_equity']/(merged_pd_banks['market_equity_bd']+merged_pd_banks['market_equity'])
merged_pd_banks[['tot_assets_rat', 'book_debt_rat', 'book_equity_rat','market_equity_rat']]
# these ratios are between 0.1 and 0.2 off of the paper's numbers for banks

tot_assets_rat  book_debt_rat  book_equity_rat  market_equity_rat
year                                                                   
1960        0.101051       0.097816         0.104998                NaN
1961        0.102634       0.104030         0.101072           1.000000
1962        0.101197       0.096436         0.100964           0.149831
1963        0.082464       0.078232         0.125286           0.137084
1964        0.118005       0.114588         0.148939           0.178940
1965        0.115581       0.112214         0.148033           0.192640
1966        0.113131       0.110196         0.148883           0.159231
1967        0.110094       0.107535         0.144982           0.153154
1968        0.109918       0.107698         0.142862           0.143166
1969        0.115196       0.113060         0.144378           0.136529
1970        0.120612       0.118340         0.153924           0.136496
1971        0.124040       0.122138         0.153444           0.150355
1972        0.124496       0.123264         0.146247           0.126651
1973        0.128958       0.128070         0.146481           0.109855
1974        0.127357       0.126467         0.142319           0.116103
1975        0.130283       0.129823         0.143140           0.118946
1976        0.132840       0.132219         0.143655           0.137518
1977        0.136991       0.136457         0.145526           0.140517
1978        0.135950       0.135116         0.152322           0.131234
1979        0.154922       0.148739         0.264677           0.345312
1980        0.156230       0.150113         0.263425           0.357121
1981        0.159632       0.153695         0.262061           0.325874
1982        0.151570       0.145602         0.252308           0.381110
1983        0.203048       0.197952         0.286735           0.392020
1984        0.237213       0.232827         0.304681           0.364778
1985        0.242605       0.238597         0.303601           0.339673
1986        0.259351       0.255599         0.318532           0.371797
1987        0.287219       0.283966         0.338834           0.365102
1988        0.287304       0.284297         0.331957           0.321760
1989        0.342217       0.340682         0.368446           0.402246
1990        0.345039       0.343960         0.357335           0.426220
1991        0.332581       0.332893         0.332370           0.399629
1992        0.299610       0.300461         0.294828           0.351351
1993        0.376936       0.380269         0.312302           0.410522
1994        0.328485       0.334525         0.244052           0.262571
1995        0.378822       0.384547         0.299486           0.277967
1996        0.375259       0.381936         0.281123           0.265318
1997        0.437567       0.445114         0.320975           0.291535
1998        0.555063       0.560891         0.462367           0.436432
1999        0.576372       0.581997         0.487877           0.535822
2000        0.565955       0.571285         0.488314           0.559386
2001        0.565640       0.571107         0.488103           0.552738
2002        0.510887       0.516092         0.435654           0.479512
2003        0.498134       0.502282         0.437781           0.498347
2004        0.561310       0.566733         0.486376           0.528794
2005        0.625073       0.630087         0.533502           0.531436
2006        0.614540       0.620975         0.514152           0.519417
2007        0.642070       0.649717         0.498303           0.500789
2008        0.648890       0.655157         0.491951           0.440916
2009        0.619405       0.622274         0.578368           0.555836
2010        0.635027       0.639117         0.576238           0.543943
2011        0.646267       0.650133         0.588606           0.522882
2012        0.633805       0.637328         0.583039           0.556373

In [179]:
tick_inv = ticks.copy()
tick_inv['End Date'] = pd.to_datetime(tick_inv['End Date'].replace({'Current': datetime.today()}))

# Now apply the filter
tick_inv_filtered = tick_inv[tick_inv['End Date'] <= '2012-12-31']
tick_inv_filtered = tick_inv_filtered[tick_inv_filtered['gvkey'] == 0]
tick_inv_filtered

Primary Dealer                      Holding Company  \
12                   Becker Paribas Inc.                       Banque Paribas   
19                Brophy, Gestal, Knight                                  NaN   
23       CARROLL MCENTEE & MCGINLEY INC.                                  NaN   
24                C.F. CHILDS & CO., INC                                  NaN   
26               CHEMICAL SECURITIES INC            Chemical Bank Corporation   
42                Drexel Burnham Lambert                                  NaN   
43                  D.W. RICH & CO., INC                                  NaN   
44               EASTBRIDGE CAPITAL INC.                                  NaN   
45                             FI Dupont                                  NaN   
50                FIRST PENNCO SEC. INC.                                  NaN   
54          HARRIS GOVERNMENT SECURITIES                                  NaN   
57               IRVING SECURITIES, INC.                 Irving Trust Company   
61    KLEINWORT BENSON  GOV'T SEC., INC.               Kleinwort Benson Group   
66                  MALON S. ANDRUS INC.                                  NaN   
76          THE NIKKO SECURITIES CO. INT           Nikko Securities Co., Ltd.   
80                NUVEEN GOV'T SEC. INC.                   Nuveen Corporation   
81                          NY Hanseatic                                  NaN   
84                   PARIBAS CORPORATION                    BNP Paribas Group   
88                    Robertson Stephens                                  NaN   
91         SANWA SECURITIES (USA) CO., L                      Sanwa Bank Ltd.   
92          SBC WARBURG DILLON READ INC.         Swiss Bank Corporation (SBC)   
93   SECOND DISTRICT SECURITIES CO., INC                                  NaN   
94                     Securities Groups                                  NaN   
98             SG COWEN SECURITIES CORP.                Société Générale S.A.   
99              S.G. WARBURG & CO., INC.               S.G. Warburg Group PLC   
104     THOMSON MCKINNON SECURITIES INC.          Thomson McKinnon Group Inc.   
107        WERTHEIM SCHRODER & CO., INC.         Schroder Wertheim & Co. Inc.   
110       YAMAICHI INT'L (AMERICA), INC.  Yamaichi Securities Company Limited   

                   Ticker  Start Date   End Date  Permco  gvkey  
12   BNP (Euronext Paris)  11/17/1971 1984-09-10       0      0  
19                    NaN    5/8/1987 1988-06-19       0      0  
23                    NaN   9/29/1976 1994-05-06       0      0  
24                    NaN   5/19/1960 1965-06-29       0      0  
26             CHL (NYSE)   5/19/1960 1996-03-31       0      0  
42                    NaN   5/19/1960 1990-03-28       0      0  
43                    NaN   5/19/1960 1969-12-31       0      0  
44                    NaN   6/18/1992 1998-05-29       0      0  
45                    NaN  12/12/1968 1973-07-18       0      0  
50                    NaN    3/7/1974 1980-08-27       0      0  
54                    NaN   7/15/1965 1995-05-31       0      0  
57                    NaN   5/19/1960 1989-07-31       0      0  
61                    NaN   2/13/1980 1989-12-27       0      0  
66                    NaN   5/19/1960 1965-11-24       0      0  
76                   8601  12/22/1987 1999-01-03       0      0  
80                    NaN  11/18/1971 1980-08-27       0      0  
81                    NaN    2/8/1984 1984-07-26       0      0  
84   BNP (Euronext Paris)    5/1/1997 2000-09-14       0      0  
88                    NaN   10/1/1997 1998-09-30       0      0  
91                    NaN   6/20/1988 1998-07-20       0      0  
92                    SBC   3/29/1990 1998-06-28       0      0  
93                    NaN   6/15/1961 1980-08-27       0      0  
94                    NaN   5/19/1960 1983-06-05       0      0  
98   GLE (Euronext Paris)    7/1/1999 2001-10-31       0      0  
99             UBS (NYSE) 

In [177]:
tick_inv_filtered[['Primary Dealer', 'Start']].tolist()

['Becker Paribas Inc.',
 'Brophy, Gestal, Knight',
 'CARROLL MCENTEE & MCGINLEY INC.',
 'C.F. CHILDS & CO., INC',
 'CHEMICAL SECURITIES INC',
 'Drexel Burnham Lambert',
 'D.W. RICH & CO., INC',
 'FI Dupont',
 'FIRST PENNCO SEC. INC.',
 'HARRIS GOVERNMENT SECURITIES',
 'IRVING SECURITIES, INC.',
 "KLEINWORT BENSON  GOV'T SEC., INC.",
 'MALON S. ANDRUS INC.',
 "NUVEEN GOV'T SEC. INC.",
 'NY Hanseatic',
 'SECOND DISTRICT SECURITIES CO., INC',
 'Securities Groups',
 'S.G. WARBURG & CO., INC.',
 'THOMSON MCKINNON SECURITIES INC.',
 'WERTHEIM SCHRODER & CO., INC.']

Creation of table 2 writeup: Table 2 was the first table our group was tasked with replicating. This table involves getting numbers such as total assets or market equity from CRSP Compustat database for different groups of firms, with the limitation that the firms be US domestic and have data available in CRSP Compustat. These groups were comprised of the primary dealers of the New York Fed, which are detailed in table A.1, all broker dealers, all banks, and all firms in Compustat. There were three time ranges that we observed, 1960-2012, 1960-1990, and 1990-2012. Replicating table 2 turned out to be a lot more involved than initially expected, due largely to the condition that we use the data of holding companies of the primary dealers rather than of the actual primary dealer. This required us to manually go through each individual dealer in the list of 112, and look up information about their holding company during a specified time period where they were considered a primary dealer. This was not a simple task because there were many cases where holding companies shifted multiple time within the specified period - forcing us to try and guess how the original author of the paper made this decision. We ended up choosing the holding company that held the primary dealer for the longest period of time during the specified interval. An alternative method would be to have multiple holding company data pulls for a single primary dealer, which may be the most accurate and most time consuming methodology. What was also a challenge was that firms seemed to have multiple names that they went by, making research and lookups of these firms difficult and easy to make an error.

After finishing the main lookup file, ticks.csv, which including all primary dealers listed in table A.1, their holding companies, their start and end dates, and their permcos and gvkeys (identifiers for CRSP and Compustat respectively) we then focused on pulling data from the WRDS database. After establishing a connection with WRDS we were able to query data from multiple sources. We made use of the CRSP Compustat historical link table to do most of this research, as well as some holding companies lookups on Google and chatGPT. We queried a full list of link table data from the WRDS website and downloaded it in Excel format to be able to filter and make notes. We also queried the link table from the WRDS historical table and used this table to help merge SIC codes with our completed ticks.csv dataset. For the financials dataset, we opted to query data by gvkey since the data for table 2 was specifically for company level financials which are found through Compustat.
  
After having the primary dealer data in the proper format of total assets, total book debt, book equity, and market equity, it was now time to create our subgroups and query data for them. The first one we worked on was the broker dealer dataset, which, as specified in the paper, included all primary dealers in our main dataset plus any firms with the SIC code 6211 or 6221. Using the historical linktable and our primary dealer linktable we created our subset and then used it to query financial data from Compustat. Now with our two different subsets of data, we merged them on the date key and then performed our ratio calculation. This calculation was the total value of primary dealers divided by the sum of the total value of primary dealers and the total value of non-primary dealer broker dealers. We then repeated this methodology for the banks comparison group which was generally comprised of firms with the SIC code 6020. We found that the bank ratios were farther from the paper than the broker dealer ratios. This could be for a multitude of reasons, potentially from differences in how holding companies were handled. 